core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -606,56 +620,73 @@ public SourceDistributionTaskSource(
+        {
+            ImmutableList.Builder<TaskDescriptor> result = ImmutableList.builder();
+            while (true) {
+                Optional<TaskDescriptor> readyTask = getReadyTask(splits, otherSplitSets, nodeRequirements);
+                if (readyTask.isEmpty()) {
+                    break;
+                }
+                result.add(readyTask.get());
             }
 
-            if (!locallyAccessibleSplitBuffer.isEmpty()) {
-                for (HostAddress remoteHost : locallyAccessibleSplitBuffer.keySet()) {
-                    List<Split> splits = ImmutableList.copyOf(locallyAccessibleSplitBuffer.get(remoteHost));
-                    if (!splits.isEmpty()) {
-                        locallyAccessibleSplitBuffer.values().forEach(otherHostSplits -> splits.forEach(otherHostSplits::remove));
-                        result.add(new TaskDescriptor(
-                                currentPartitionId++,
-                                ImmutableListMultimap.<PlanNodeId, Split>builder().putAll(partitionedSourceNodeId, splits).build(),
-                                replicatedExchangeSourceHandles,
-                                new NodeRequirements(catalogRequirement, ImmutableSet.of(remoteHost))));
-                    }
+            if (includeRemainder && !splits.isEmpty()) {
+                result.add(buildTaskDescriptor(splits, nodeRequirements));
+                for (Set<Split> otherSplits : otherSplitSets) {
+                    otherSplits.removeAll(splits);
                 }
-                locallyAccessibleSplitBuffer.clear();
+                splits.clear();
             }
-
             return result.build();
         }
 
-        private static <T> List<T> removeN(Collection<T> collection, int n)
+        private Optional<TaskDescriptor> getReadyTask(Set<Split> splits, List<Set<Split>> otherSplitSets, NodeRequirements nodeRequirements)
         {
-            ImmutableList.Builder<T> result = ImmutableList.builder();
-            Iterator<T> iterator = collection.iterator();
-            for (int i = 0; i < n && iterator.hasNext(); i++) {
-                T item = iterator.next();
-                iterator.remove();
-                result.add(item);
+            ImmutableList.Builder<Split> chosenSplitsBuilder = ImmutableList.builder();
+            int splitCount = 0;
+            int totalSlitWeight = 0;
+            for (Split split : splits) {
+                totalSlitWeight += split.getSplitWeight().getRawValue();
+                splitCount++;
+                chosenSplitsBuilder.add(split);
+
+                if (totalSlitWeight >= targetPartitionSplitWeight && splitCount >= minPartitionSplitCount || splitCount >= maxPartitionSplitCount) {
+                    ImmutableList<Split> chosenSplits = chosenSplitsBuilder.build();
+                    for (Set<Split> otherSplits : otherSplitSets) {
+                        chosenSplits.forEach(otherSplits::remove);
```

Comment:
Could this be `otherSplits.removeAll(chosenSplits);` ?

Reply:
Used to be that. But IntelliJ compains that `removeAll` can be slow sometimes (not sure why) and suggest this form. So I changed.

Reply:
Ah yeah, looks like the implementation in AbstractMap can be quadratic depending on the size of the set relative to the events to remove.

Reply:
Yeah - looks like it should manifest with slowness only if the collection passed as an argument to `removeAll` has slow `contains()`, which is not the case here. But this statement is readable too imo.